greenriver/hmis-warehouse
# 185708758 hmis api synchronization

## Description
https://www.pivotaltracker.com/story/show/185708758
https://github.com/greenriver/hmis-frontend/pull/470

Proactively handle conflicts that could arise from multiple people acting on shared records. Adds rails optimistic locking to some HMIS HUD client, customAssessment, and enrollment models to protect us conflicting mutations. 
* Conflicts result in an error message shown to the client instructing them to reload the page
* optimistic locking is disabled for the warehouse models to prevent unintended side-effects
* This is not intended to be comprehensive and more work is needed, perhaps expanding locking to other record types

Other changes:
* lock around the enrollment on submit_assessment, enrollment id is now required
* wrap submit_form mutation in transaction to ensure updates are atomic
* submit_household_assessment takes an array of {id, lockVersion} inputs rather than just the ID

## Testing / QA
Intake / Exit Assessment:
- open two tabs with the same assessment
- click "save changes" in each tab. One of the tabs should succeed and the other one should display an error message

Delete Assessment:
- open two tabs with the same assessment
- update the assessment in one tab, in the other tab attempt to delete the assessment, the delete should fail

Update Client Details:
- open two tabs with the same client
- click "save changes" in each tab. One of the tabs should succeed and the other one should display an error message
- click "submit" in each tab. One of the tabs should succeed and the other one should display an error message

Delete Client:
- open two tabs with the same client
- update the client details in one tab, in the other tab attempt to delete the client, the delete should fail

## Type of change


## Checklist before requesting review
